Buildings/Fluid/BuriedPipes/GroundCoupling.mo

```diff
@@ -0,0 +1,283 @@
+within Buildings.Fluid.BuriedPipes;
+model GroundCoupling "Thermal coupling between buried pipes and ground"
+  parameter Integer nPip(min=1) "Number of buried pipes";
+
+  replaceable parameter Buildings.HeatTransfer.Data.Soil.Generic soiDat
+    "Soil thermal properties";
+  replaceable parameter
+    Buildings.BoundaryConditions.GroundTemperature.ClimaticConstants.Generic
+    cliCon "Surface temperature climatic conditions";
+
+  parameter Integer nSeg(min=1) = 1
+    "Number of axial segments (and number of heat ports for each pipe)";
+
+  parameter Modelica.SIunits.Length len[nSeg] "Pipes length";
+
+  parameter Modelica.SIunits.Length dep[nPip] "Pipes buried depth";
+  parameter Modelica.SIunits.Length pos[nPip]
+    "Pipes horizontal coordinate (to an arbitrary reference point)";
+  parameter Modelica.SIunits.Length rad[nPip] "Pipes external radius";
+
+  Modelica.Thermal.HeatTransfer.Interfaces.HeatPort_a ports[nPip, nSeg]
+    "Buried pipes heatports"
+    annotation (Placement(transformation(extent={{-10,-10},{10,10}},
+        origin={0,-100})));
+  Buildings.BoundaryConditions.GroundTemperature.UndisturbedSoilTemperature
+    soi(cliCon=cliCon, soiDat=soiDat, dep=depMea) "Soil temperature";
+
+protected
+  parameter Modelica.SIunits.Length depMea = sum(dep) / nPip "Average depth";
+  parameter Real P[nPip,nPip]=BaseClasses.groundCouplingFactors(
+      nPip,
+      dep,
+      pos,
+      rad) "Thermal coupling geometric factors";
+
+equation
+  for seg in 1:nSeg loop
+    ports[:,seg].T .- soi.port.T = P * ports[:,seg].Q_flow /
+        (2 * Modelica.Constants.pi * soiDat.k * len[seg]);
```

Comment:
Just to confirm, the paper you wanted to link was [Dynamic equation-based thermo-hydraulic pipe model for district heating and cooling systems](https://www.sciencedirect.com/science/article/pii/S0196890417307975) (not sure how this happened but your link is just a few digits off, but amazingly, equation (3) in the paper you linked is somewhat relevant).

Regarding your observation, I am not sure I fully understand. In particular, when you say:
> In this equation ports[:,seg].T corresponds to the temperature at the exterior surface of the pipe, which is the variable the equation is solved for (ports[:,seg].Q_flow being computed by the instance of PlugFlowPipe this component is connected to).

I do not understand why `PlugFlowPipe` is solving for `Q` and `GroundCoupling` is solving for `T_b` (to reprise the notation from Van Der Heijde). My understanding is that the models don't solve for one or the other, but simply define the relationships between `T_b`, `Q` and other variables, some of which are known.
Put differently, if I were to flatten out the two models into one, I would pretty much end up with a system of three equations:
```
(1) T_out = T_b + (T_in - T_b) * E
(2) Q  = (T_out - T_in) * C
(3) Q = (T_b - T_g) * P
```
where `E`, `C` and `P` are constant terms (assuming constant mass flow rate). Equations (1) and (2) are from the `PlugFlowPipe` model and (3) is from  `GroundCoupling` (for a single pipe).
The way I understand your comment, I would just be shuffling the equations around, but in essence, would keep the same system:
```
(1) T_b = (T_out - T_in) * E / (1 - E)
(2) T_out = T_in + Q / C
(3) Q = (T_b - T_g) * P
```
Am I getting that right? If so, I don't fully understand why this would be interpreted differently par the solver.

Reply:
You are right to consider the system of equations in its whole and I realize that my comment is confusing.
Let me try again.

1. The "contract" behind the `PlugFlowPipe` model is that a temperature boundary condition can be provided that is decoupled from the fluid temperature at the time scale at which the integration is done, that is the transport delay along the pipe. Based on that, the model is able to provide an analytical formulation of the local heat flow rate, integrate it and compute the fluid outlet temperature. 
2. The `GroundCoupling` model is based on the average pipe-soil interface temperature. Based on that, the model provides an analytical formulation of the average heat flow rate.

It seems to me that coupling those two models yields an ambiguous computation scheme where the contract 1. is no more satisfied. But that contract is also no more needed since we are able to compute the average heat flow rate from 2. So what we really need 1. to provide is rather a formulation of the mean surface temperature, knowing the average heat flow rate and fluid temperature variation along the pipe.